src/query/sql/src/planner/optimizer/property/selectivity.rs

```diff
@@ -378,3 +381,58 @@ fn update_statistic(
     }
     Ok(())
 }
+
+fn evaluate_equal_by_histogram(const_datum: &Datum, col_stat: &ColumnStat) -> Result<f64> {
+    let hist = col_stat.histogram.as_ref().unwrap();
+    let min = col_stat.min.to_double()?;
+    let max = col_stat.max.to_double()?;
+    let mut range = 0.0;
+    for (idx, bucket) in hist.buckets_iter().enumerate() {
+        if idx == 0 {
+            continue;
+        }
+        let bucket_min = hist.buckets[idx - 1].upper_bound().to_double()?;
+        let bucket_max = bucket.upper_bound().to_double()?;
+        // If the bucket max is less than min, skip it
+        if bucket_max < min {
+            continue;
+        }
+        // If the bucket min is greater than max, stop iteration
+        if bucket_min >= max {
+            break;
+        }
+        if (bucket_min > min && bucket_max <= max) || bucket_max == bucket_min {
+            // ---min---bucket_min---bucket_max---max---
+            range += 1.0;
+        } else if bucket_min <= min {
+            // left part
+            range += (bucket_max - min) / (bucket_max - bucket_min);
+        } else if bucket_max > max {
+            // right part
+            range += (max - bucket_min) / (bucket_max - bucket_min);
+        }
+    }
+
+    let mut equal_range = 0.0;
+    let const_value = const_datum.to_double()?;
+    for (idx, bucket) in hist.buckets_iter().enumerate() {
+        if idx == 0 {
+            continue;
+        }
+        let bucket_min = hist.buckets[idx - 1].upper_bound().to_double()?;
+        let bucket_max = bucket.upper_bound().to_double()?;
+        // If the bucket max is less than min, skip it
+        if bucket_max < const_value || bucket.num_distinct() == 0.0 {
```

Comment:
`bucket_max <= const_value`, because the bucket range is [bucket_min, bucket_max).